src/topojson/topojson.ml

```diff
@@ -1 +1,351 @@
+                  let geo = J.to_list (decode_or_err base_of_json) list in
+                  Result.map (fun v -> (Collection v, fm)) geo
+              | None ->
+                  Error
+                    (`Msg
+                      "A geometry collection should have a member called \
+                       geometries"))
+          | Ok typ -> Error (`Msg ("Unknown type of geometry " ^ typ))
+          | Error _ as e -> e)
+      | None ->
+          Error
+            (`Msg
+              "A Geojson text should contain one object with a member `type`.")
+
+    let rec to_json ?bbox = function
+      | Point point, foreign_members ->
+          Point.to_json ?bbox ~foreign_members point
+      | MultiPoint mp, foreign_members ->
+          MultiPoint.to_json ?bbox ~foreign_members mp
+      | LineString ls, foreign_members ->
+          LineString.to_json ?bbox ~foreign_members ls
+      | MultiLineString mls, foreign_members ->
+          MultiLineString.to_json ?bbox ~foreign_members mls
+      | Polygon p, foreign_members -> Polygon.to_json ?bbox ~foreign_members p
+      | MultiPolygon mp, foreign_members ->
+          MultiPolygon.to_json ?bbox ~foreign_members mp
+      | Collection c, foreign_members ->
+          J.obj
+            ([
+               ("type", J.string "GeometryCollection");
+               ("geometries", J.list to_json c);
+             ]
+            @ bbox_to_json_or_empty bbox
+            @ foreign_members)
+
+    let foreign_members (_, fm) = fm
+  end
+
+  module Topology = struct
+    type json = J.t
+
+    let keys_in_use = [ "type"; "arcs"; "objects"; "transform"; "bbox" ]
+
+    let foreign_members json =
+      match J.to_obj json with
+      | Ok assoc ->
+          List.filter (fun (k, _v) -> not (List.mem k keys_in_use)) assoc
+      | Error _ -> []
+
+    module Position = struct
+      type t = float array
+
+      let lng t = t.(0)
+      let lat t = t.(1)
+      let altitude t = try Some t.(2) with _ -> None
+
+      let v ?altitude ~lng ~lat () =
+        match altitude with
+        | Some f -> [| lng; lat; f |]
+        | None -> [| lng; lat |]
+
+      let equal l1 l2 =
+        let n1 = Array.length l1 and n2 = Array.length l2 in
+        if n1 <> n2 then false
+        else
+          let rec loop i =
+            if i = n1 then true
+            else if Float.equal (Array.unsafe_get l1 i) (Array.unsafe_get l2 i)
+            then loop (succ i)
+            else false
+          in
+          loop 0
+
+      let to_json arr = J.array J.float arr
+    end
+
+    type t = { objects : (string * json) list; arcs : Position.t array array }
+
+    let of_json json =
+      match (J.find json [ "objects" ], J.find json [ "arcs" ]) with
+      | Some objects, Some arcs ->
+          let* objects = J.to_obj objects in
+          let* arcs =
+            J.to_array
+              (decode_or_err (J.to_array (decode_or_err J.to_float)))
+              arcs
+          in
+          { objects; arcs }
```

Comment:
```suggestion
          Ok { objects; arcs }
```

If your final let-statement isn't a `let+` then you'll need to wrap this in an `Ok`.